### PR TITLE
DWPDAT-141: Implement Pagination Service

### PR DESCRIPTION
**Pull request by AbubakrBhula**

### CODE REVIEW

**Comment by hmrc-web-operations** [comment]
Thank you for submitting a pull request on this HMRC repository. A member of the HMRC team will review your changes and action accordingly.

HMRC reviewer: please reference [this runbook](https://github.com/hmrc/build-and-deploy/blob/main/runbooks/reviewing_external_pull_requests.md) for guidance on how to proceed.

File: app/uk/gov/hmrc/app/benefitEligibility/service/PaginationService.scala [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    contributionAndCreditsPaging: Option[ContributionAndCreditsPaging]
+)
+
+final case class PaginationResult(
+    benefitType: BenefitType,
+    liabilitiesResult: List[LiabilityResult],
+    marriageDetailsResult: Option[MarriageDetailsResult],
+    contributionCreditResult: ContributionCreditPagingResult,
+    benefitSchemeMembershipDetailsData: Option[BenefitSchemeMembershipDetailsData]
+) {
```
Comment (lthompsonjr) [inline app/uk/gov/hmrc/app/benefitEligibility/service/PaginationService.scala]: move these classes to a separate file

File: app/uk/gov/hmrc/app/benefitEligibility/service/PaginationService.scala [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+            liabilitiesResult = liabilitiesResult,
+            marriageDetailsResult = marriageDetailsResult,
+            contributionCreditResult = ContributionCreditPagingResult(contributionCreditResult, existingPageTask.contributionAndCreditsPaging),
+            benefitSchemeMembershipDetailsData = benefitSchemeMembershipDetailsData
+          )
+          PageTask
+            .continuePaginatingTask(paginationResult)
+            .map { newPageTask =>
+              pageTaskRepo.upsert(newPageTask).flatMap(id => pageTaskRepo.delete(existingPageTask.id))
+            }
```
Comment (lthompsonjr) [inline app/uk/gov/hmrc/app/benefitEligibility/service/PaginationService.scala]: we shouldn't need to use delete if we're using upsert

File: app/uk/gov/hmrc/app/benefitEligibility/service/PaginationService.scala [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    benefitSchemeMembershipDetailsData: Option[BenefitSchemeMembershipDetailsData]
+) {
+
+  private def shouldPage: Boolean =
+    List(
+      PaginationSource.fromLiabilities(liabilitiesResult),
+      PaginationSource.fromBenefitSchemeMembershipDetails(benefitSchemeMembershipDetailsData),
+      PaginationSource.fromMarriageDetails(marriageDetailsResult)
+    ).flatten.nonEmpty || contributionCreditResult.contributionAndCreditsPaging.exists(_.niContributionAndCreditsTaxWindows.tail.nonEmpty)
+  lazy val nextCursor: Option[PaginationCursor] = if (shouldPage) Some(PaginationCursor(UUID.randomUUID())) else None
```
Comment (lthompsonjr) [inline app/uk/gov/hmrc/app/benefitEligibility/service/PaginationService.scala]: add new line

File: app/uk/gov/hmrc/app/benefitEligibility/service/PaginationService.scala [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+  private[service] def liabilitiesConnectorFetchData(
+      pageTask: PageTask
+  )(implicit headerCarrier: HeaderCarrier): EitherT[Future, BenefitEligibilityError, List[LiabilityResult]] =
+    pageTask.liabilitiesPaging.map { page =>
+      liabilitySummaryDetailsConnector
+        .fetchData(pageTask.benefitType, page.callBackURL.toString)
+    }.sequence
+
+  private[service] def marriageDetailsConnectorFetchData(pageTask: PageTask)(
```
Comment (lthompsonjr) [inline app/uk/gov/hmrc/app/benefitEligibility/service/PaginationService.scala]: change name to fetchMarriageDetails

File: app/uk/gov/hmrc/app/benefitEligibility/service/PaginationService.scala [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+                    .flatMap(i => EitherT.pure[Future, BenefitEligibilityError]((detailsResult, i)))
+                case None => EitherT.pure[Future, BenefitEligibilityError]((detailsResult, Nil))
+              }
+
+          }
+      }
+      .sequence
+      .map(maybeTupledResult => maybeTupledResult.map(tuple => BenefitSchemeMembershipDetailsData(tuple._1, tuple._2)))
+
+  private[service] def creditsAndContributionsFetchData(pageTask: PageTask, nationInsuranceNumber: Identifier)(
```
Comment (lthompsonjr) [inline app/uk/gov/hmrc/app/benefitEligibility/service/PaginationService.scala]: change name to fetchCreditsAndContributionsData

File: app/uk/gov/hmrc/app/benefitEligibility/service/PaginationService.scala [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+            .continuePaginatingTask(paginationResult)
+            .map { newPageTask =>
+              pageTaskRepo.upsert(newPageTask).flatMap(id => pageTaskRepo.delete(existingPageTask.id))
+            }
+            .sequence
+            .map(_ => paginationResult)
+      }
+    }
+
+  private[service] def liabilitiesConnectorFetchData(
```
Comment (lthompsonjr) [inline app/uk/gov/hmrc/app/benefitEligibility/service/PaginationService.scala]: chane name to fetchLiabilitiesData

File: app/uk/gov/hmrc/app/benefitEligibility/service/PaginationService.scala [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+class PaginationService @Inject() (
+    liabilitySummaryDetailsConnector: LiabilitySummaryDetailsConnector,
+    niContributionsAndCreditsConnector: NiContributionsAndCreditsConnector,
+    marriageDetailsConnector: MarriageDetailsConnector,
+    schemeMembershipDetailsConnector: SchemeMembershipDetailsConnector,
+    benefitSchemeDetailsConnector: BenefitSchemeDetailsConnector,
+    pageTaskRepo: BenefitEligibilityRepository
+)(implicit ec: ExecutionContext) {
+
+  def setPagingTask(pageTask: PageTask): EitherT[Future, BenefitEligibilityError, UUID] =
```
Comment (lthompsonjr) [inline app/uk/gov/hmrc/app/benefitEligibility/service/PaginationService.scala]: change to addTask

File: app/uk/gov/hmrc/app/benefitEligibility/service/PaginationService.scala
Comment (lthompsonjr) [inline app/uk/gov/hmrc/app/benefitEligibility/service/PaginationService.scala:1]: Add some logging in this service as well

File: test/uk/gov/hmrc/app/benefitEligibility/service/PaginationServiceSpec.scala
Comment (lthompsonjr) [inline test/uk/gov/hmrc/app/benefitEligibility/service/PaginationServiceSpec.scala:1]: You need to add test for the failure cases
for example if there's a database failure or a call to a connector fails with a random fault.

File: test/uk/gov/hmrc/app/nationalinsurancecontributionandcreditsapi/controllers/NICCControllerSpec.scala
Comment (lthompsonjr) [inline test/uk/gov/hmrc/app/nationalinsurancecontributionandcreditsapi/controllers/NICCControllerSpec.scala:1]: Revert this file, you shouldn't need to change anything here

File: app/uk/gov/hmrc/app/benefitEligibility/model/response/BenefitEligibilityResponse.scala [hunk 8]
```diff
@@ -57,6 +61,111 @@ sealed trait BenefitEligibilityInfoResponse { self =>
...
+    maybeContributionCreditResult match {
+      case Some(result) =>
+        result match {
+          case NpsApiResult.FailureResult(apiName, result) => NiContributionsAndCreditsSuccessResponse(None, None, None)
+          case NpsApiResult.SuccessResult(apiName, successResponse) => successResponse
+        }
+      case None => NiContributionsAndCreditsSuccessResponse(None, None, None)
+    }
+
+  def from(
```
Comment (lthompsonjr) [inline app/uk/gov/hmrc/app/benefitEligibility/model/response/BenefitEligibilityResponse.scala:81]: You need to add tests for this

File: app/uk/gov/hmrc/app/benefitEligibility/model/response/BenefitEligibilityResponse.scala [hunk 9]
```diff
@@ -57,6 +61,111 @@ sealed trait BenefitEligibilityInfoResponse { self =>
...
+              niContributionsAndCreditsResult =
+                toContributionCreditResult(paginationResult.contributionCreditResult.contributionCreditResult),
+              paginationResult.nextCursor
+            )
+          )
+
+        // TODO Abu change this to match what's being done in MA and GYSP
+        // also create private methods for the logic under each case statement to tidy this up
+        // it actually might be better to create a separate helper that houses some of these methods as this file is getting quite big
+        case BenefitType.BSP =>
```
Comment (lthompsonjr) [inline app/uk/gov/hmrc/app/benefitEligibility/model/response/BenefitEligibilityResponse.scala]: This still needs to be done.

File: app/uk/gov/hmrc/app/benefitEligibility/repository/BenefitEligibilityRepository.scala [hunk 10]
```diff
@@ -52,10 +58,43 @@ class BenefitEligibilityRepositoryImpl @Inject() (mongoComponent: MongoComponent
...
+    collection
+      .deleteOne(Filters.equal("id", id.toString))
+      .toFuture()
+      .attemptT
+      .map {
+        _.wasAcknowledged()
+      }
+      .leftMap(error => DatabaseError(error))
+
+  def upsert(pageTask: PageTask): EitherT[Future, BenefitEligibilityError, UUID] = {
```
Comment (lthompsonjr) [inline app/uk/gov/hmrc/app/benefitEligibility/repository/BenefitEligibilityRepository.scala]: you need to add tests for this

File: it/test/uk/gov/hmrc/app/benefitEligibility/repository/BenefitEligibilityDataRepositoryItSpec.scala
Comment (lthompsonjr) [inline it/test/uk/gov/hmrc/app/benefitEligibility/repository/BenefitEligibilityDataRepositoryItSpec.scala:1]: you also need tests for negative scenarios

**Comment by lthompsonjr** [comment]
Add tests for the GYSP BSP and MA services for when pagination is required

File: app/uk/gov/hmrc/app/benefitEligibility/service/BereavementSupportPaymentDataRetrievalService.scala [hunk 11]
```diff
@@ -55,13 +62,47 @@ class BereavementSupportPaymentDataRetrievalService @Inject() (
...
+                benefitType = BSP,
+                benefitSchemeMembershipDetailsPaging = None,
+                marriageDetailsPaging = marriageDetailsPaginate,
+                liabilitiesPaging = Nil,
+                contributionAndCreditsPaging = Some(niContributionsCreditsPaginate)
+              )
+            )
+            .map(id => result.copy(nextCursor = Some(PaginationCursor(id))))
+        } else {
+          println("ABU ")
```
Comment (lthompsonjr) [inline app/uk/gov/hmrc/app/benefitEligibility/service/BereavementSupportPaymentDataRetrievalService.scala]: remove

File: test/uk/gov/hmrc/app/benefitEligibility/service/PaginationServiceSpec.scala [hunk 12]
```diff
@@ -0,0 +1,580 @@
...
+    liabilitySummaryDetailsConnector = mockLiabilitySummaryDetailsConnector,
+    niContributionsAndCreditsConnector = mockNiContributionsAndCreditsConnector,
+    marriageDetailsConnector = mockMarriageDetailsConnector,
+    schemeMembershipDetailsConnector = mockSchemeMembershipDetailsConnector,
+    benefitSchemeDetailsConnector = mockBenefitSchemeDetailsConnector,
+    pageTaskRepo = mockBenefitEligibilityRepository
+  )
+
+  "PaginationService" - {
+    ".setPagingTask" - {
```
Comment (lthompsonjr) [inline test/uk/gov/hmrc/app/benefitEligibility/service/PaginationServiceSpec.scala]: this is now addTask

File: app/uk/gov/hmrc/app/benefitEligibility/service/PaginationService.scala
Comment (lthompsonjr) [inline app/uk/gov/hmrc/app/benefitEligibility/service/PaginationService.scala:1]: Add integration tests for this as well `PaginationServiceItSpec`

File: it/test/uk/gov/hmrc/app/benefitEligibility/repository/BenefitEligibilityDataRepositoryItSpec.scala [hunk 13]
```diff
@@ -55,123 +58,171 @@ class BenefitEligibilityDataRepositoryItSpec
...
 
         pageTasksList.foreach(insert)
 
         val pageTasks = Table("page_task", pageTasksList: _*)
 
-        forAll(pageTasks)(pageTaskList => repository.getItem(pageTaskList.id).futureValue mustBe Some(pageTaskList))
+        forAll(pageTasks)(pageTask => repository.getItem(pageTask.id).value.futureValue shouldBe Right(pageTask))
       }
-      "should return None when a nonexistent id is parsed" in {
+      "Should fail and return DataBase error" in {
```
Comment (lthompsonjr) [inline it/test/uk/gov/hmrc/app/benefitEligibility/repository/BenefitEligibilityDataRepositoryItSpec.scala]: change to something like "Should return Database error if there is an unexpected db failure"

File: it/test/uk/gov/hmrc/app/benefitEligibility/repository/BenefitEligibilityDataRepositoryItSpec.scala [hunk 14]
```diff
@@ -55,123 +58,171 @@ class BenefitEligibilityDataRepositoryItSpec
...
-        val uuidTwo: UUID   = UUID.fromString("fa356ed8-27f2-4c62-8204-386366713356")
-        val uuidThree: UUID = UUID.fromString("f2968e2a-37cd-4f4e-9d66-bb0351c6dd6c")
+        repository.getItem(uuidTwo.value).value.futureValue shouldBe Left(DatabaseError(RecordNotFound(uuidTwo.value)))
+      }
+    }
 
-        val pageListForIdOne   = List(Pages(Class2MAReceipts, "SomeCallBackURLOne"))
-        val pageListForIdTwo   = List(Pages(Liabilities, "SomeCallBackURLTwo"))
-        val pageListForIdThree = List(Pages(MarriageDetails, "SomeCallBackURLThree"))
+    ".delete" - {
```
Comment (lthompsonjr) [inline it/test/uk/gov/hmrc/app/benefitEligibility/repository/BenefitEligibilityDataRepositoryItSpec.scala]: missing failure test case

File: it/test/uk/gov/hmrc/app/benefitEligibility/repository/BenefitEligibilityDataRepositoryItSpec.scala [hunk 15]
```diff
@@ -55,123 +58,171 @@ class BenefitEligibilityDataRepositoryItSpec
...
 
-        forAll(pageTasks)(pageTaskList =>
-          repository.getItem(UUID.fromString("ca4eab1b-55fa-4d6a-8aa4-7eb3debc6db7")).futureValue mustBe None
-        )
+        forAll(pageTasks)(pageTaskList => repository.delete(uuidTwo.value).value.futureValue.isRight)
+        findAll().futureValue must contain theSameElementsAs List(maPageTask, gyspPageTask)
       }
     }
+    ".upsert" - {
+      "Should return a UUID for BSP" in {
```
Comment (lthompsonjr) [inline it/test/uk/gov/hmrc/app/benefitEligibility/repository/BenefitEligibilityDataRepositoryItSpec.scala]: this should be `"should insert a new BspPageTask"` 
you also need a test for `"should overwrite an existing BspPageTask"`

File: it/test/uk/gov/hmrc/app/benefitEligibility/repository/BenefitEligibilityDataRepositoryItSpec.scala [hunk 16]
```diff
@@ -55,123 +58,171 @@ class BenefitEligibilityDataRepositoryItSpec
...
 
-    ".delete" - {
-      "should successfully delete an item by ID" in {
-
-        val uuidOne: UUID   = UUID.fromString("54c99a34-86d9-4154-b617-5f60c7064bde")
-        val uuidTwo: UUID   = UUID.fromString("fa356ed8-27f2-4c62-8204-386366713356")
-        val uuidThree: UUID = UUID.fromString("f2968e2a-37cd-4f4e-9d66-bb0351c6dd6c")
+        repository.upsert(bspPageTask).value.futureValue shouldBe Right(uuidOne.value)
+      }
+      "Should return a UUID for MA" in {
```
Comment (lthompsonjr) [inline it/test/uk/gov/hmrc/app/benefitEligibility/repository/BenefitEligibilityDataRepositoryItSpec.scala]: same as 
https://github.com/hmrc/national-insurance-contribution-and-credits-api/pull/104/changes#r2905237591

File: it/test/uk/gov/hmrc/app/benefitEligibility/repository/BenefitEligibilityDataRepositoryItSpec.scala [hunk 17]
```diff
@@ -55,123 +58,171 @@ class BenefitEligibilityDataRepositoryItSpec
...
-            pageListForIdTwo,
-            GYSP
-          ),
-          PageTask(
-            uuidThree,
-            pageListForIdThree,
-            BSP
+        repository.upsert(maPageTask).value.futureValue shouldBe Right(uuidOne.value)
+      }
+      "Should return a UUID for GYSP" in {
```
Comment (lthompsonjr) [inline it/test/uk/gov/hmrc/app/benefitEligibility/repository/BenefitEligibilityDataRepositoryItSpec.scala]: same as 
https://github.com/hmrc/national-insurance-contribution-and-credits-api/pull/104/changes#r2905237591

**Review by lthompsonjr — APPROVED** [review]
